docs/plugin/errors.md

```diff
@@ -0,0 +1,514 @@
+        {
+            _parent.Deactivated(connectionId);
+        }
+
+        BEGIN_INTERFACE_MAP(Notification)
+        INTERFACE_ENTRY(RPC::IRemoteConnection::INotification)
+        END_INTERFACE_MAP
+
+    private:
+        TestPlugin& _parent;
+    };
+
+public:
+    TestPlugin()
+        : _connectionId(0)
+        , _service(nullptr)
+        , _testPlugin(nullptr)
+        , _notification(this)
+    {
+    }
+    ~TestPlugin() override = default;
+
+    TestPlugin(TestPlugin&&) = delete;
+    TestPlugin(const TestPlugin&) = delete;
+    TestPlugin& operator=(TestPlugin&&) = delete;
+    TestPlugin& operator=(const TestPlugin&) = delete;
+
+    BEGIN_INTERFACE_MAP(TestPlugin)
+    INTERFACE_ENTRY(PluginHost::IPlugin)
+    INTERFACE_ENTRY(PluginHost::IDispatcher)
+    INTERFACE_AGGREGATE(Exchange::ITestPlugin, _testPlugin)
+    END_INTERFACE_MAP
+
+public:
+    // IPlugin methods
+    const string Initialize(PluginHost::IShell* service) override;
+    void Deinitialize(PluginHost::IShell* service) override;
+    string Information() const override;
+
+private:
+    void Deactivated(RPC::IRemoteConnection* connection);
+
+private:
+    uint32_t _connectionId;
+    PluginHost::IShell* _service;
+    Exchange::ITestPlugin* _testPlugin;
+    Core::Sink<Notification> _notification;
+};
+```
+
+
+```c++ linenums="1" title="TestPlugin.cpp" hl_lines="12 32 61-72"
+const string TestPlugin::Initialize(PluginHost::IShell* service)
+{
+    ASSERT(_service == nullptr);
+    ASSERT(_connectionId == 0);
+
+    string result = {};
+
+    _service = service;
+    _service->AddRef();
+
+    // Register for COM-RPC connection/disconnection notifications
+    _service->Register(&_notification);
+
+    // Instantiate the ITestPlugin interface (which will spawn the OOP side if running in OOP mode)
+    // Store connection ID in _connectionId
+    _testPlugin = _service->Root<Exchange::ITestPlugin>(_connectionId, 2000, _T("TestPluginImplementation"));
+
+    if (!_testPlugin) {
+        // Error occurred, return non-empty string
+        result = "Failed to create ITestPlugin";
+    }
+
+    return result;
+}
+
+void TestPlugin::Deinitialize(PluginHost::IShell* service)
+{
+    if (service != nullptr) {
+        ASSERT(_service == service);
+
+        // Unsubscribe from connection notification first to prevent any false-positives
+        _service->Unregister(&_notification);
+
+        if (_testPlugin != nullptr) {
+            // This should release the last reference and destruct the object. If not,
+            // there's something else holding on to it and we have a leak
+            VARIABLE_IS_NOT_USED uint32_t result = _testPlugin->Release();
+            ASSERT(result == Core::ERROR_DESTRUCTION_SUCCEEDED);
+            _testPlugin = nullptr;
+
+            // Shut down the out-of-process connection if still running
+            RPC::IRemoteConnection* connection(_service->RemoteConnection(_connectionId));
```

Comment:
This needs a small correction, this line needs to move to above line 255. The release on the _testPlugin could also release the connection and then you would not be able to get it here anymore. If you do it before the release on _testPlugin it will always be available

Reply:
Corrected